voidcrew/code/modules/overmap/dynamic_events.dm

```diff
@@ -63,74 +42,25 @@
-		DYNAMIC_WORLD_ASTEROID = 30)
-
-	if(force_encounter)
-		chosen = force_encounter
-	else
-		chosen = pickweight(probabilities)
-	mass = rand(50, 100) * 1000000 //50 to 100 million tonnes //this was a stupid feature
-	switch(chosen)
-		if(DYNAMIC_WORLD_LAVA)
-			name = "strange lava planet"
-			desc = "A very weak energy signal originating from a planet with lots of seismic and volcanic activity."
-			planet = DYNAMIC_WORLD_LAVA
-			icon_state = "globe"
-			color = COLOR_ORANGE
-		if(DYNAMIC_WORLD_ICE)
-			name = "strange ice planet"
-			desc = "A very weak energy signal originating from a planet with traces of water and extremely low temperatures."
-			planet = DYNAMIC_WORLD_ICE
-			icon_state = "globe"
-			color = COLOR_BLUE_LIGHT
-		if(DYNAMIC_WORLD_JUNGLE)
-			name = "strange jungle planet"
-			desc = "A very weak energy signal originating from a planet teeming with life."
-			planet = DYNAMIC_WORLD_JUNGLE
-			icon_state = "globe"
-			color = COLOR_LIME
-		if(DYNAMIC_WORLD_SAND)
-			name = "strange sand planet"
-			desc = "A very weak energy signal originating from a planet with many traces of silica."
-			planet = DYNAMIC_WORLD_SAND
-			icon_state = "globe"
-			color = COLOR_GRAY
-		if(DYNAMIC_WORLD_ROCKPLANET)
-			name = "strange rock planet"
-			desc = "A very weak energy signal originating from a abandoned industrial planet."
-			planet = DYNAMIC_WORLD_ROCKPLANET
-			icon_state = "globe"
-			color = COLOR_BROWN
-		if(DYNAMIC_WORLD_REEBE)
-			name = "???"
-			desc = "Some sort of strange portal. Theres no identification of what this is."
-			planet = DYNAMIC_WORLD_REEBE
-			icon_state = "wormhole"
-			color = COLOR_YELLOW
-		if(DYNAMIC_WORLD_ASTEROID)
-			name = "large asteroid"
-			desc = "A large asteroid with significant traces of minerals."
-			planet = DYNAMIC_WORLD_ASTEROID
-			icon_state = "asteroid"
-			mass = rand(1, 1000) * 100
-			color = COLOR_GRAY
-		if(DYNAMIC_WORLD_SPACERUIN)
-			name = "weak energy signal"
-			desc = "A very weak energy signal emenating from space."
-			planet = DYNAMIC_WORLD_SPACERUIN
-			icon_state = "strange_event"
-			color = null
-			mass = 0 //Space doesn't weigh anything
-	desc += !preserve_level && "It may not still be here if you leave it."
+	if(isnull(planet))
+		generate_probabilites()
+		planet = pickweight(probabilities)
+
+	var/datum/overmap/planet/temp_planet = new planet
+	name = temp_planet.name
+	desc = temp_planet.desc + "[preserve_level ? "" : " It may not still be here if you leave it."]"
+	icon_state = temp_planet.icon_state
+	color = temp_planet.color
+	qdel(temp_planet)
+
+/obj/structure/overmap/dynamic/proc/generate_probabilites()
+	if (!isnull(probabilities))
+		return
+	probabilities = list()
+	for (var/path in subtypesof(/datum/overmap/planet))
+		var/datum/overmap/planet/temp_planet = new path
+		probabilities |= list(temp_planet.type = min(length(temp_planet.ruin_list), temp_planet.spawn_rate))
```

Comment:
Why not just temp_planet.spawn_rate? Planets with low ruin amount be extremely rare.

Reply:
I'm not sure, this is just how shiptest did it. I'll try that out and compare